webextension/scripts/bulk-save.js

```diff
@@ -0,0 +1,170 @@
+// bulk-save.js
+
+// from 'utils.js'
+/*   global isNotExcludedUrl, isValidUrl, openByWindowSetting, checkAuthentication, wmAvailabilityCheck, hostURL */
+
+let urlList = new Set()
+let newSetLength = 0
+let oldSetLength = 0
+let totalUrlCount = 0
+
+// check if the user is Logged in, if not, prompt to Login
+checkAuthentication((result) => {
+  if (result && result.message && result.message === 'You need to be logged in to use Save Page Now.') {
+    $('#save-all-box').addClass('flip-inside')
+    $('#start-bulk-save').attr('disabled', true)
+    $('#login-to-save').click(() => {
+      openByWindowSetting('https://archive.org/account/login')
+    })
+  }
+})
+
+function displayList(list) {
+  newSetLength = list.size
+  if (newSetLength > oldSetLength) {
+    for (let item of Array.from([...list]).slice(oldSetLength, newSetLength)) {
+      let row = $('<div class="url-list flex-container">')
+      let del = $('<div class="delete-btn">').text('X')
+      let span = $('<div class="url-item">').append(item)
+
+      $('#list-container').append(
+        row.append(del, span)
+      )
+    }
+    oldSetLength = newSetLength
+  }
+}
+
+function processNode(node) {
+  // recursively process child nodes
+  if (node.children) {
+    node.children.forEach((child) => { processNode(child) })
+  }
+
+  // access leaf nodes: Bookmarked URLs
+  if (node.url) {
+    if (isValidUrl(node.url) && isNotExcludedUrl(node.url)) {
+      urlList.add(node.url)
+      displayList(urlList)
+    }
+  }
+}
+
+// import all bookmarked URLs
+function importBookmarks() {
+  $('#empty-list-err').hide()
+  chrome.bookmarks.getTree((itemTree) => {
```

Comment:
I got this error after tapping "Import All Bookmarks":
`Uncaught TypeError: Cannot read property 'getTree' of undefined`

Reply:
Ok, thank you. I will remove this error.

Reply:
I'm not sure why, as it later worked correctly. Although I was logged out when this occurred, then was logged in when it worked correctly, but I wouldn't think that would matter when trying to read the bookmarks list. Or maybe I just forgot to reload the extension initially and it needed to refresh the manifest file to gain access to bookmarks. (since new permission was added.)

Reply:
I got the error again, but then refreshed the extension and it worked. Good idea to at least check that `chrome.bookmarks` isn't null.

Reply:
Thank you. Fixed.